beacon-chain/p2p/peers/score_block_providers.go

```diff
@@ -0,0 +1,278 @@
+	// requested and returned, but none have been processed yet. Allows distinguishing between
+	// peers that haven't yet returned anything and peers that returned garbage.
+	SlowProcessedBlocksPenalty float64
+	// DecayInterval defines how often requested/returned/processed stats should be decayed.
+	DecayInterval time.Duration
+	// Decay specifies the factor (must be < 1.0) by which block provider's stats is decayed.
+	Decay float64
+}
+
+// newBlockProviderScorer creates block provider scoring service.
+func newBlockProviderScorer(
+	ctx context.Context, store *peerDataStore, config *BlockProviderScorerConfig) *BlockProviderScorer {
+	if config == nil {
+		config = &BlockProviderScorerConfig{}
+	}
+	scorer := &BlockProviderScorer{
+		ctx:    ctx,
+		config: config,
+		store:  store,
+	}
+	if scorer.config.StartScore == 0.0 {
+		scorer.config.StartScore = DefaultBlockProviderStartScore
+	}
+	if scorer.config.ReturnedBlocksWeight == 0.0 {
+		scorer.config.ReturnedBlocksWeight = DefaultBlockProviderReturnedBlocksWeight
+	}
+	if scorer.config.SlowReturnedBlocksPenalty == 0.0 {
+		scorer.config.SlowReturnedBlocksPenalty = DefaultSlowReturnedBlocksPenalty
+	}
+	if scorer.config.ProcessedBlocksWeight == 0.0 {
+		scorer.config.ProcessedBlocksWeight = DefaultBlockProviderProcessedBlocksWeight
+	}
+	if scorer.config.SlowProcessedBlocksPenalty == 0.0 {
+		scorer.config.SlowProcessedBlocksPenalty = DefaultSlowProcessedBlocksPenalty
+	}
+	if scorer.config.DecayInterval == 0 {
+		scorer.config.DecayInterval = DefaultBlockProviderDecayInterval
+	}
+	if scorer.config.Decay == 0.0 {
+		scorer.config.Decay = DefaultBlockProviderDecay
+	}
+	return scorer
+}
+
+// Score calculates and returns total score based on returned and processed blocks.
+func (s *BlockProviderScorer) Score(pid peer.ID) float64 {
+	s.store.RLock()
+	defer s.store.RUnlock()
+	return s.score(pid)
+}
+
+// score is a lock-free version of Score.
+func (s *BlockProviderScorer) score(pid peer.ID) float64 {
+	score := s.Params().StartScore
+	peerData, ok := s.store.peers[pid]
+	if ok && peerData.requestedBlocks > 0 {
+		// Score returned/requested ratio.
+		returnedBlocksScore := float64(peerData.returnedBlocks) / float64(peerData.requestedBlocks)
+		returnedBlocksScore = returnedBlocksScore * s.config.ReturnedBlocksWeight
+		score += returnedBlocksScore
+
+		if peerData.requestedBlocks > peerData.returnedBlocks {
+			factor := float64(peerData.requestedBlocks-peerData.returnedBlocks) / float64(peerData.requestedBlocks)
+			score += s.config.SlowReturnedBlocksPenalty * factor
+		}
+
+		// Score processed/requested ratio.
+		processedBlocksScore := float64(peerData.processedBlocks) / float64(peerData.requestedBlocks)
+		processedBlocksScore = processedBlocksScore * s.config.ProcessedBlocksWeight
+		score += processedBlocksScore
+
+		if peerData.returnedBlocks > 0 && peerData.returnedBlocks > peerData.processedBlocks {
```

Comment:
this is dangerous, there are many possible reasons we cannot process a block. What if for the batch that we receive from a peer , we do not have a valid parent ? We would unfairly penalise that peer, you could have a bad peer constantly causing other peers to be descored.

Reply:
As block providers scores are to relatively score the performance of a peers (not banning them, but to understand how their performance is compared to other peers when selecting the next peer to fetch data from), if our `init-sync` algorithm allows single bad peer constantly throw away data from valid peers, then it is not a problem in scoring, it is a problem in syncing algorithm as it is open to such an attack vector. And we need to devise a way for good data to be utilized, not thrown away (btw, that bad peer shouldn't be able to escape descoring too).

Reply:
>  if our init-sync algorithm allows single bad peer constantly throw away data from valid peers, then it is not a problem in scoring, it is a problem in syncing algorithm as it is open to such an attack vector. And we need to devise a way for good data to be utilized, not thrown away (btw, that bad peer shouldn't be able to escape descoring too.

I guess the main goal would be to prevent peers like that from making us throw out our whole batch and penalise that peer, wouldnt doing something like this just make the problem worse ?